// knip/dist/typescript/resolveModuleNames.js
import { existsSync } from 'node:fs';
import ts from 'typescript';
import { sanitizeSpecifier } from '../util/modules.js';
import { dirname, extname, isInternal, join } from '../util/path.js';
import { ensureRealFilePath, isVirtualFilePath } from './utils.js';
export function createCustomModuleResolver(customSys, compilerOptions, virtualFileExtensions) {
    function resolveModuleNames(moduleNames, containingFile) {
        return moduleNames.map(moduleName => resolveModuleName(moduleName, containingFile));
    }
    function resolveModuleName(name, containingFile) {
        const sanitizedSpecifier = sanitizeSpecifier(name);
        const tsResolvedModule = ts.resolveModuleName(sanitizedSpecifier, containingFile, compilerOptions, ts.sys).resolvedModule;
        if (!tsResolvedModule) {
            const extension = extname(sanitizedSpecifier);
            if (extension && isInternal(sanitizedSpecifier) && !virtualFileExtensions.includes(extension)) {
                const resolvedFileName = join(dirname(containingFile), sanitizedSpecifier);
                if (existsSync(resolvedFileName)) {
                    return {
                        resolvedFileName,
                        extension,
                        isExternalLibraryImport: false,
                        resolvedUsingTsExtension: false,
                    };
                }
            }
        }
        if (virtualFileExtensions.length === 0)
            return tsResolvedModule;
        if (tsResolvedModule && !isVirtualFilePath(tsResolvedModule.resolvedFileName, virtualFileExtensions)) {
            return tsResolvedModule;
        }
        const customResolvedModule = ts.resolveModuleName(name, containingFile, compilerOptions, customSys).resolvedModule;
        if (!customResolvedModule || !isVirtualFilePath(customResolvedModule.resolvedFileName, virtualFileExtensions)) {
            return customResolvedModule;
        }
        const resolvedFileName = ensureRealFilePath(customResolvedModule.resolvedFileName, virtualFileExtensions);
        const resolvedModule = {
            extension: ts.Extension.Js,
            resolvedFileName,
            isExternalLibraryImport: customResolvedModule.isExternalLibraryImport,
        };
        return resolvedModule;
    }
    return resolveModuleNames;
}
