// knip/dist/plugins/gatsby/index.js
import { timerify } from '../../util/Performance.js';
import { hasDependency, load } from '../../util/plugin.js';
import { toProductionEntryPattern } from '../../util/protocols.js';
export const NAME = 'Gatsby';
export const ENABLERS = ['gatsby', 'gatsby-cli'];
export const isEnabled = ({ dependencies }) => hasDependency(dependencies, ENABLERS);
export const CONFIG_FILE_PATTERNS = [
    'gatsby-{config,node}.{js,jsx,ts,tsx}',
    'plugins/**/gatsby-node.{js,jsx,ts,tsx}',
];
export const PRODUCTION_ENTRY_FILE_PATTERNS = [
    'gatsby-{browser,ssr}.{js,jsx,ts,tsx}',
    'src/api/**/*.{js,ts}',
    'src/pages/**/*.{js,jsx,ts,tsx}',
    'src/templates/**/*.{js,jsx,ts,tsx}',
    'src/html.{js,jsx,ts,tsx}',
    'plugins/**/gatsby-{browser,ssr}.{js,jsx,ts,tsx}',
];
const findGatsbyDependencies = async (configFilePath, { isProduction }) => {
    const config = await load(configFilePath);
    const entryPatterns = PRODUCTION_ENTRY_FILE_PATTERNS.map(toProductionEntryPattern);
    if (isProduction)
        return entryPatterns;
    if (/gatsby-config/.test(configFilePath)) {
        return config.plugins.map(plugin => (typeof plugin === 'string' ? plugin : plugin.resolve));
    }
    if (/gatsby-node/.test(configFilePath)) {
        const plugins = new Set();
        const actions = { setBabelPlugin: plugin => plugins.add(plugin.name) };
        const _config = config;
        if (typeof _config.onCreateBabelConfig === 'function') {
            _config.onCreateBabelConfig({ actions });
        }
        return Array.from(plugins);
    }
    return [];
};
export const findDependencies = timerify(findGatsbyDependencies);
